src/backends/all.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::prelude::*;
+use anyhow::Result;
+
+macro_rules! append {
+    ($($backend:ident),*) => {
+        pub fn append(&mut self, other: &mut Self) {
+            $(
+                self.$backend.append(&mut other.$backend);
+            )*
+        }
+    };
+}
+macro_rules! is_empty {
+    ($($backend:ident),*) => {
+        pub fn is_empty(&self) ->bool {
+            $(
+                self.$backend.is_empty() &&
+            )* true
+        }
+    };
+}
+macro_rules! to_package_ids {
+    ($($backend:ident),*) => {
+        pub fn to_package_ids(&self) -> PackageIds {
+            PackageIds {
+                $(
+                    $backend: self.$backend.keys().cloned().collect(),
+                )*
+            }
+        }
+    };
+}
+
+macro_rules! x {
```

Comment:
Can probably use a better name here.

Reply:
The reasoning behind the name `x` is that the macro is designed to be called only once but anonymous macros are not possible in rust yet so I had to do it this way, the advantage being that if the content of the macro changes such that the current name was no longer accurate the macro name would not drift out of correctness. But I don't mind splitting out the macro and naming them if you think it would be nicer.

Reply:
It's the same for lots of the other macros used for running per-backend code due to the different backend package types.

Reply:
Fair enough, but I think a proper name and splitting would give two benefits:
- Naming it would give context to what the macro is doing
- Splitting it would make it easy to inspect the generated code by using rust-analyzer's macro expansion feature.

Reply:
Good points, I'll split and name the big macro.